File: examples/chained_io.m
:- module chained_io.
:- interface.
:- import_module io.

:- pred main(io::di, io::uo) is det.

:- implementation.
:- import_module string.

    % Demonstrates that Mercury io.write_string calls chain through the
    % HLDS conjunction sequence.  The bridge sees three GoalCalls in a
    % GoalConj and emits a sequence of print_str calls; the !IO state
    % thread is discarded (the runtime is side-effectful regardless).
main(!IO) :-
    io.write_string("first\n", !IO),
    io.write_string("second\n", !IO),
    io.write_string("third\n", !IO).

File: examples/chained_io.rs
// Demonstrates that three sequential Rust println! calls work through
// the Frankenstein bridge.  Each println! expands to
//   Arguments::<'_>::from_str → elided
//   std::io::_print → print_str
// and the three calls become three consecutive let-bound print_str
// applications in Frankenstein Core.

pub fn main() {
    println!("first");
    println!("second");
    println!("third");
}
